### ISSUE TITLE
Begin resolving `cargo-semver-checks` blockers for merging into cargo

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust-project-goals#104]
| Metadata      | |
| --------      | --- |
| Owner(s)      | @obi1kenobi |
| Team(s)       | [cargo](https://github.com/rust-lang/cargo) |
| Goal document | [2024h2/cargo-semver-checks](https://rust-lang.github.io/rust-project-goals/2024h2/cargo-semver-checks.html) |

## Summary

Design and implement `cargo-semver-checks` functionality that lies on the critical path for merging the tool into cargo itself.

## Tasks and status

* [ ] Implementation of cargo manifest linting + CLI (@obi1kenobi)
* [ ] Initial design for cross-crate checking (@obi1kenobi)
* [ ] Initial design for type-checking lints (@obi1kenobi)
* [ ] Discussion and moral support ([cargo](https://github.com/rust-lang/cargo) ![Team][])

[Team]: https://img.shields.io/badge/Team%20ask-red

**Comment by nikomatsakis** [comment]
This issue is intended for status updates only.

For general questions or comments, please contact the owner(s) directly.

**Comment by obi1kenobi** [comment]
Status update as of September 1.

**Key developments:** We've laid a lot of the groundwork to make manifest linting possible.
- In preparation for the significant expansion of our CLI, we now have the ability to test how CLI invocations get interpreted internally. We can also now snapshot the output of any CLI invocation over a given workspace, allowing us to proceed with refactoring the implementation without breaking existing use cases.
- We have a concrete design for the expansion of the CLI, and for the Trustfall schema changes that will be necessary to enable manifest linting.

Not blocked on anything. No opportunities to contribute _yet_, but stay tuned — there will be some in future updates!

**Comment by obi1kenobi** [comment]
No major updates to report. Preparing my talk for next week's EuroRust has taken away most of my free time.

**Comment by obi1kenobi** [comment]
**Key developments:**
- Partial schema design and implementation of type information in lints
- Resolved multi-team questions that were blocking cross-crate checking
- Refactor of cargo-semver-checks data handling to enable supporting manifest information

### Partial schema design and implementation of type information in lints

The schema we use for writing lints now contains information on:
- [generic parameters (lifetime, type, or const)](https://github.com/obi1kenobi/trustfall-rustdoc-adapter/blob/7063ac96da3d78f656061a09443e657eb16ae41b/src/rustdoc_schema.graphql#L2265-L2363) on structs, enums, unions, traits, and impl blocks
- [trait bounds on generic type parameters](https://github.com/obi1kenobi/trustfall-rustdoc-adapter/blob/7063ac96da3d78f656061a09443e657eb16ae41b/src/rustdoc_schema.graphql#L2310-L2321)

This newly allows the implementation of breaking-change lints such as "generic const parameter no longer has default value" or "trait bound removed from generic parameter in trait method."

Additionally, @suaviloquence has implemented the ability to print rustdoc JSON type information into human-readable source code form, which will allow us to produce higher-quality diagnostics for type-related breaking changes. This required a lot of care, perseverance, and testing — thank you 🙏

### Resolved multi-team questions that were blocking cross-crate checking

For a long time, cross-crate checking was blocked on a multi-team coordination problem: cargo-semver-checks needs more data from rustdoc, which in turn needs more data from rustc. This was a classic "everyone's third-most important thing" problem: important, but not critical and therefore never top of the agenda for resolving the open questions there.

In-person meetings with Rust Project members during my trips to RustConf and EuroRust helped nudge this forward:
- The [compiler team MCP](https://github.com/rust-lang/compiler-team/issues/635) intended to provide additional data to rustdoc was merged and can now be implemented.
- I synced up with T-rustdoc members on how rustdoc would then present that data to cargo-semver-checks in rustdoc JSON. Along the way, we were able to discuss and agree on several additional improvements to rustdoc JSON that improved performance in cargo-semver-checks.

### Refactor of cargo-semver-checks data handling

Plumbing in manifest information was proving challenging due to a significant amount of tech debt accrued in the cargo-semver-checks code that loads the data of scanned crates. [This PR](https://github.com/obi1kenobi/cargo-semver-checks/pull/1001) is a significant clean-up and refactor of that portion of the tool, allowing us to move forward with retrieving manifest information while also supporting other quality-of-life improvements.

With the mountain of tech debt largely resolved, I expect that adding manifest information should be _relatively_ straightforward from here.

This portion of the goal also greatly benefited from the Google Summer of Code work that @suaviloquence did — it wouldn't have been possible without the _significant_ effort he put into improving our test suites and harnesses.

**Comment by obi1kenobi** [comment]
**Key developments:**
- Support for cargo manifest linting is now merged: https://github.com/obi1kenobi/cargo-semver-checks/pull/1007

### Cargo manifest linting

As a recap, the idea here is catching breakage caused by _manifest_ (`Cargo.toml`) changes, not just source code changes.

An example of such breakage is the removal of a package feature: any crates that enabled the removed feature will no longer build.

The PR above builds upon the data-handling refactor from the previous update post, and makes it possible to lint manifests for breaking changes. The PR includes the `feature_missing` lint that checks for the exact case mentioned above. That is but one of many new lints that are now possible with this feature. Hooray! 🎉

**Comment by obi1kenobi** [comment]
Thanks, Niko! Copying in the new tasks from the 2025h1 period:
* [ ] Prototype cross-crate linting using workarounds (@obi1kenobi)
* [ ] Allow linting generic types, lifetimes, bounds (@obi1kenobi)
* [ ] Handle "special cases" like `'static` and `?Sized` (@obi1kenobi)
* [ ] Handle `#[doc(hidden)]` in sealed trait analysis (@obi1kenobi)
* [ ] Discussion and moral support ([cargo](https://github.com/rust-lang/cargo), [rustdoc]

**Comment by obi1kenobi** [comment]
**Key developments:**
- Sealed trait analysis correctly handles `#[doc(hidden)]` items. This completes one checkbox of this goal!
- We shipped a series of lints detecting breakage in generic types, lifetimes, and const generics. One of them has already caught accidental breakage in the real world!

[`cargo-semver-checks` v0.40](https://github.com/obi1kenobi/cargo-semver-checks/releases/tag/v0.40.0), released today, includes a variety of improvements to sealed trait analysis. They can be summarized as "smarter, faster, more correct," and will have an immediate positive impact on popular crates such as `diesel` and `zerocopy`.

While we [already shipped a series of lints](https://github.com/obi1kenobi/cargo-semver-checks/releases/tag/v0.39.0) detecting generics-related breakage, more work is needed to complete that checkbox. This, and the "special cases like `'static` and `?Sized`", will be the focus of upcoming work.

**Comment by obi1kenobi** [comment]
TL;DR:

Earlier this month, we completed one checkbox of the goal: `#[doc(hidden)]` in sealed trait analysis, live in `cargo-semver-checks` v0.40. We also made significant progress on type system modeling, which is part of two more checkboxes.
- We shipped method receiver types in our schema, enabling more than a dozen new lints.
- We have a draft schema for `?Sized` bounds, and are putting the finishing touches on `'static` and "outlives" bounds. More lints will follow here.
- We also have a draft schema for the new `use<>` precise capturing syntax.

Additionally, `cargo-semver-checks` is participating in Google Summer of Code, so this month we had the privilege of merging many contributions from new contributors who are considering applying for GSoC with us! We're looking forward to this summer, and would like to wish the candidates good luck in the application process!

**Comment by obi1kenobi** [comment]
We encountered some speedbumps this month.

TL;DR:
- While working on `'static` and "outlives" bounds, we discovered Rust's ability to _imply_ bounds that are not stated explicitly at the definition site.
- Implied bounds are load-bearing for SemVer; failure to take them into account will produce _both_ false-positives _and_ false-negatives.
- While technical limitations make it infeasible for `cargo-semver-checks` to correctly deduce implied bounds, `rustc` has this capability internally.
- We have [asked the rustdoc team](https://rust-lang.zulipchat.com/#narrow/channel/266220-t-rustdoc/topic/Show.20implied.20bounds.20in.20rustdoc.20JSON/near/515146429) to expose implied bounds in rustdoc JSON by using those `rustc` internal APIs.

There are some good news as well, though! While looking at the `#[target_feature]` attribute:
- We discovered previously-undocumented SemVer hazards.
- We [discovered a case of unsoundness](https://github.com/rust-lang/rust/issues/139368) when that attribute is used on trait methods.
- With the help of contributors and the rustdoc team, rustdoc JSON began including additional information that will help future `cargo-semver-checks` versions catch those SemVer hazards.

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by obi1kenobi** [comment]
Just removed the duplicate posts, guessing from a script that had a bad day.

**Comment by obi1kenobi** [comment]
Status update as of November 1

**Key developments:**
- Draft PR for exposing implied bounds in rustdoc JSON: https://github.com/rust-lang/rust/pull/148379
- A concrete plan for how that new info turns into dozens of new lints covering many kinds of bounds

Linting `?Sized` and `'static` bounds turned out to be quite a bit more complex than I anticipated. The key issue is that seeing `T: Foo + ?Sized` does not guarantee that `T` can be unsized, since we might have `Foo: Sized` which renders the `?Sized` relaxation ineffective. Similarly, seeing `T: Foo` might also non-obviously imply `T: 'static` via a similar implied bound.

Failure to correctly account for implied bounds would lead to catastrophic false-positives and false-negatives. For example, changing `T: Foo` to `T: Foo + 'static` could be a major breaking change or a no-op, depending on whether we have `Foo: 'static` (either directly or implicitly via other trait bounds).

We cannot determine implied bounds using information present in rustdoc JSON today, so the rustdoc team and I have been iterating on the best way to compute and include that information in rustdoc JSON. Assuming something similar to [the aforementioned PR](https://github.com/rust-lang/rust/pull/148379) becomes part of rustdoc JSON, `cargo-semver-checks` stands to gain several dozen new lints covering these tricky cases over trait associated types, generic type parameters, and APIT/RPIT/RPITIT.

**Comment by obi1kenobi** [comment]
Google Summer of Code 2025 is complete + finally some movement on cross-crate linting! 🚀

**Key developments**
- Two students had a successful conclusion of Google Summer of Code working on `cargo-semver-checks` — [find more](https://blog.rust-lang.org/2025/11/18/gsoc-2025-results/#make-cargo-semver-checks-faster) [details here](https://blog.rust-lang.org/2025/11/18/gsoc-2025-results/#enable-witness-generation-in-cargo-semver-checks)!
- rustdoc JSON now includes rlib information, following the design for cross-crate rustdoc JSON info created at RustWeek 2025: https://github.com/rust-lang/rust/pull/149043
- A cargo issue was discovered that prevents this rlib info from being used; it's currently being triaged: https://github.com/rust-lang/cargo/issues/16291
- Once that's resolved, we'll have enough here for a basic prototype. Getting features right in dependencies will likely require more work due to having many more cargo-related edge cases.

**Comment by tomassedovic** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2026

**Comment by obi1kenobi** [comment]
There's been significant internal progress. As I described in [my year-end overview post](https://predr.ag/blog/cargo-semver-checks-2025-year-in-review/#the-path-forward-for-2026-and-beyond), we're aiming to solve `cargo-semver-checks`' key limitations, which requires solving several hard problems at once.

We're about 9000 lines of code deep into implementing a normalized type representation, a prerequisite for efficient type-checking lints. The gist of the idea is:
- We want to detect when types may have changed, so we can trigger a type-checking run.
- That type-checking run is fairly expensive: we need to generate a witness program, and run `cargo check` on it up to twice.
- Rust has many ways to make syntactic changes to types that are semantically irrelevant. For example, consider renaming all `'a` lifetimes to `'b` — this shouldn't run type-checking, because nothing _semantic_ changed so there cannot be a Semantic Versioning violation.

We expect we're a handful of PRs away from having a suitable implementation of this normalization process. Our focus will then pivot toward cross-crate rustdoc JSON use, which will both address a number of false-positives and false-negatives per se, and is another prerequisite for type-checking lints.

Engaging with these hard problems is also producing benefits for Rustaceans today. First, digging deeply into the codebase helps discover bugs and edge cases, and then it produces a perfect opportunity to address them. For example, here are a couple of fixes that will be part of the next `cargo-semver-checks` release:
- Better handling of edge cases around `#[doc(hidden)]` applied to glob re-exports: https://github.com/obi1kenobi/trustfall-rustdoc-adapter/pull/1072
- Bug fixes for accidental duplication of lint results: https://github.com/obi1kenobi/trustfall-rustdoc-adapter/pull/1066

**Comment by obi1kenobi** [comment]
At the request of Rust project members, we took a brief detour to implement breakage-checking for the Rust standard library. This work [is very close to](https://github.com/rust-lang/rust/pull/159671) [the finish line](https://github.com/rust-lang/rust/pull/160253) and involved about 15kLoC across 30+ PRs, largely in order to handle stability attributes correctly. I plan to publish a blog post on it when it's complete.

Like in the previous post, this process also caused us to discover and fix pre-existing issues in `cargo-semver-checks` affecting users today. For example, two lints had [existing false-positives fixed](https://github.com/obi1kenobi/cargo-semver-checks/pull/1685) as a result of us reasoning through the (much more complex) edge cases of partial stability in the Rust standard library.

Additionally, we also improved the type normalization code to be more robust to generic parameter additions or removals. Previously, all generic parameters of the same kind (types, lifetimes, consts) shared the same normalization namespace. This caused issues. By way of analogy:
```rust
impl<A> Foo<A> {
    fn bar<B>(_: B) {}
}
```
If `Foo` gains another generic type parameter, that new parameter would get normalized to `B` instead. This would change the normalization of `bar<B>` into `bar<C>`, making it normalize differently ("appear to change"). This would have triggered many unnecessary witness checks and would have been tediously slow. With the latest fixes, [generics on a parent scope no longer change the normalization of inner scopes' generics](https://github.com/obi1kenobi/trustfall-rustdoc-adapter/pull/1088), avoiding this problem.
